auxx/src/Lang/Name.hs

```diff
@@ -0,0 +1,54 @@
+module Lang.Name
+       ( Letter(getLetter)
+       , unsafeMkLetter
+       , Name(..)
+       , unsafeMkName
+       ) where
+
+import           Prelude                           (Show (..))
+import           Universum
+
+import           Data.Char                         (isAlpha)
+import           Data.Coerce                       (coerce)
+import qualified Data.List.NonEmpty                as NonEmpty
+import           Data.List.Split                   (splitWhen)
+import qualified Data.Text.Buildable               as Buildable
+import           Test.QuickCheck.Arbitrary.Generic (Arbitrary (..), genericArbitrary,
+                                                    genericShrink)
+import           Test.QuickCheck.Gen               (suchThat)
+
+-- | Invariant: @isAlpha . getLetter = const True@
+newtype Letter = Letter { getLetter :: Char }
+    deriving (Eq, Ord, Show)
+
+unsafeMkLetter :: Char -> Letter
+unsafeMkLetter = Letter
+
+instance Arbitrary Letter where
+    arbitrary = Letter <$> arbitrary `suchThat` isAlpha
+
+newtype Name = Name (NonEmpty (NonEmpty Letter))
```

Comment:
Why do we need such representation? Why do we treat `patak-bardaq` as `[patak, bardaq]` and not just `patak-bardaq`?

Reply:
it guarantees the correct structure by code, not by a comment, so we get generic methods (i.e. `genericShrink`)

Reply:
I am not sure you got me right.
My question is: suppose we had `Name = Name (NonEmpty Letter)`, would it make some existing functionality impossible? Suppose `Letter` allowed `-`, then `patak-bardaq` would be `Name "patak-bardaq"` (pseudo-code), would it be essentially different from `Name ["patak", "bardaq"]` (which we have now) somewhere?

Reply:
Then we'd get names that start with a hyphen, end with a hyphen, and contain multiple hyphens sequentially

Reply:
Is it bad? :)

Reply:
I'd like to have more free lexical space for future additions, so broadly speaking yes. For instance `--` could be used for comments or decrement